### PR TITLE
Unify API for stanzas

### PR DESCRIPTION
**Pull request by chrzaszcz**
The main goal is to make `mongoose_stanza_api`:
- Contain all API calls needed by the GraphQL and REST API for stanzas.
- Return `{ok, Result}` or `{ErrorReasonAtom, ErrorMsg}` like the remaining API modules.
- Handle common errors for REST and GraphQL.

**Functional changes**
- Allow stanza without 'from' when 'user' is known, because 'from' is set to the user JID later in the routing process anyway.
- Return Stanza ID for GraphQL and REST. MAM ID was used for GraphQL before. We can add MAM ID as a separate GraphQL field when needed.
- Set Stanza ID if the stanza does not already have one - this is similar to the behaviour for messages.
- Do not treat `0` and `<<>>` values as undefined - this was a bit unexpected, and now in the unified API the user can always omit them when needed.
- Set correct `max_result_limit` - previously it was set to 1.
- Check for `mod_mam_pm` before querying the last messages with GraphQL. This is not done for REST, as we don't have a simple way to do it. We are not developing REST further anyway.

**Refactoring notes**
- The concept of folding over several steps is used in `mongoose_stanza_api`. We could reuse it in other modules as well (by moving to a separate module later on), but I don't want to force it. Anyway, the `fold` function, which has all the logic, is very simple. The advantages are steps reusability (even between API's in the future), and readability (the happy path is easy to follow).

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/145553) / small_tests / 689ad91f8abcbe059e6cbdc138b43bc53e1fe89b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/145553/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/145553/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-19_14.33.34/lib.mongooseim.logs/run.2022-10-19_14.33.34/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/145654) / small_tests / 163bde25bcc59eca09e5b5f8bc097490ba4dbed0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/145654/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/145654/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-20_17.10.23/lib.mongooseim.logs/run.2022-10-20_17.10.23/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/145706) / small_tests / d1f6c81c1af1e2c3b5c8b4e802d0b9814d3bf891
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/145706/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/145706/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-21_07.43.48/lib.mongooseim.logs/run.2022-10-21_07.43.48/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/145733) / small_tests / 48a5bba95cd685e42844b995598e474a03f58a24
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/145733/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/145733/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-21_08.02.26/lib.mongooseim.logs/run.2022-10-21_08.02.26/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/145806) / small_tests / 8cca66299d776451b482707c18e0c9ba798b7e8c
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/145806/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/145806/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-21_09.32.05/lib.mongooseim.logs/run.2022-10-21_09.32.06/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/145992) / small_tests / 4bcc1d3416481fe28ca596ad5a2384d6f9e8b587
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/145992/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/145992/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-21_10.43.18/lib.mongooseim.logs/run.2022-10-21_10.43.18/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/146146) / small_tests / f78be93aee8c21c09a2f544fa1a623b986577b2a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/146146/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/146146/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-21_12.38.24/lib.mongooseim.logs/run.2022-10-21_12.38.24/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/146170) / small_tests / 27bcfdd3038c4d31db785508608a09db69f2ec09
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/146170/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/146170/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-21_12.44.56/lib.mongooseim.logs/run.2022-10-21_12.44.57/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/146196) / small_tests / e27870eabc54390082f2337ed2fc915bc5d270d8
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/146196/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/146196/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-21_13.09.45/lib.mongooseim.logs/run.2022-10-21_13.09.46/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/146221) / small_tests / a29930dfddfeb18f877852762fa416919f532f91
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/146221/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/146221/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-21_13.56.05/lib.mongooseim.logs/run.2022-10-21_13.56.05/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/146247) / small_tests / 4c81efb684ebc0d41ffcc1ccb0c46895552a41b7
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3814/146247/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3814/146247/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-21_14.22.28/lib.mongooseim.logs/run.2022-10-21_14.22.28/suite.log.html)

File: src/mongoose_stanza_api.erl [hunk 1]
```diff
@@ -1,39 +1,187 @@
...
+    #xmlel{name = Name, children = [#xmlcdata{content = Text}]}.
+
+add_id(Attrs) ->
+    [{<<"id">>, mongoose_bin:gen_from_crypto()} | Attrs].
+
+-spec get_id(exml:element()) -> binary() | undefined.
+get_id(Stanza) ->
+    exml_query:attr(Stanza, <<"id">>, undefined).
+
+fold({_, _} = Result, _) ->
```
Comment (JanuszJakubiec) [inline src/mongoose_stanza_api.erl:184]: I really enjoyed reading this fragment of code 😄

**Review by JanuszJakubiec — APPROVED** [review]
Looks good

**Pull request merged**